### PR TITLE
Bug 1875247: Add package check for the channel head querying

### PR DESCRIPTION
**Pull request by jianzhangbjz**
**Description of the change:**

OLM supports multi packages in a channel, so OLM should check the head bundle based on the package. Fix bug: https://bugzilla.redhat.com/show_bug.cgi?id=1875247
Any comments are welcome! Thanks!

**Motivation for the change:**
Supports multi packages for the channel head querying

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage 
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/docs` 
- [ ] Commit messages sensible and descriptive

### CODE REVIEW

File: pkg/controller/registry/resolver/resolver.go
Comment (kevinrizza) [inline pkg/controller/registry/resolver/resolver.go:563]: nit: can this error message be `head of channel not found`?

File: pkg/controller/registry/resolver/resolver.go
Comment (kevinrizza) [inline pkg/controller/registry/resolver/resolver.go:566]: Is this change just for the case where `headCandidates == nil`? Is it necessary since above we are just already returning if that's the case?

File: pkg/controller/registry/resolver/resolver.go
Comment (jianzhangbjz) [reply pkg/controller/registry/resolver/resolver.go:563]: Sure, thanks!

File: pkg/controller/registry/resolver/resolver.go
Comment (jianzhangbjz) [reply pkg/controller/registry/resolver/resolver.go:566]: Not really, this change is for the situation of the `len(headCandidates) > 1`.  If there are multi packages in a channel, there will multi `headCandidates`, return all of them here.

**Comment by jianzhangbjz** [comment]
/retitle Bug 1875247: Add package check for the channel head querying

File: pkg/controller/registry/resolver/resolver.go
Comment (benluddy) [inline pkg/controller/registry/resolver/resolver.go:563]: This would only fire if two _consecutive_ candidates were from the same package (`[A A B]` vs. `[A B A]`), which I don't think is the intention.

**Review by benluddy — CHANGES_REQUESTED** [review]
Thanks for opening this PR.

I don't think we should be checking at runtime for a condition that should never occur -- that is, the arguments to `sortChannel` should all be in the same package and channel.

Improving the errors that are returned when there are more or fewer than 1 head candidates would be great, though. Do you mind updating this change so that it only affects the errors that are returned?

**Comment by ecordell** [comment]
I think most of this was addressed with #1735, do we want to keep this open to address error messages / etc?

File: pkg/controller/registry/resolver/resolver.go
Comment (jianzhangbjz) [reply pkg/controller/registry/resolver/resolver.go]: @benluddy Thanks for your review and suggestions! I agreed. Update the error logs.

**Comment by jianzhangbjz** [comment]
@ecordell @benluddy Thanks! I removed other parts and updated the error logs output.

File: pkg/controller/registry/resolver/resolver.go
Comment (jianzhangbjz) [reply pkg/controller/registry/resolver/resolver.go:562]: Ok, got it. I misunderstood `v.sourceInfo.Channel` and `v.sourceInfo.Package`. I thought they provides the names.
```console
entry_id    channel_na  package_name                    operatorbundle_name             replaces    depth     
----------  ----------  ------------------------------  ------------------------------  ----------  ----------
...     
9           quay-v3.3   quay-bridge-operator            quay-bridge-operator.v3.3.1     10          0         
10          quay-v3.3   quay-bridge-operator            quay-bridge-operator.v3.3.0                 1         
15          quay-v3.3   quay-operator                   red-hat-quay.v3.3.1             16          0         
16          quay-v3.3   quay-operator                   red-hat-quay.v3.3.0                         1         
17          quay-v3.3   container-security-operator     container-security-operator.v3  18          0         
18          quay-v3.3   container-security-operator     container-security-operator.v3              1 
```

**Comment by benluddy** [comment]
/retest

Looks good to me. Thanks for the contribution! This is just waiting for a second review and for the master branch to unfreeze (should happen in about a week).

**Review by ecordell — APPROVED** [review]
/lgtm

**Pull request merged**
